src/OrbitGl/TimelineTicks.cpp

```diff
@@ -69,6 +69,23 @@ std::vector<uint64_t> TimelineTicks::GetMajorTicks(uint64_t start_ns, uint64_t e
   return major_ticks;
 }
 
+int TimelineTicks::GetTimestampsNumDigitsPrecision(uint64_t start_ns, uint64_t end_ns) const {
+  constexpr int kMaxDigitsPrecision = 9;  // 1ns = 0.000'000'001s
+
+  uint64_t current_precision_ns = kNanosecondsPerSecond;
+  for (int num_digits = 0; num_digits < kMaxDigitsPrecision;
+       num_digits++, current_precision_ns /= 10) {
+    bool precision_is_enough = true;
+    for (uint64_t timestamp : GetMajorTicks(start_ns, end_ns)) {
+      if (timestamp % current_precision_ns != 0) {
+        precision_is_enough = false;
+      }
+    }
+    if (precision_is_enough) return num_digits;
+  }
+  return kMaxDigitsPrecision;
```

Comment:
This is calling `GetMajorTicks` more often than needed.

What about using a logarithm instead:

```cpp
int max_number_of_decimal_places = 0;

for (uint64_t timestamp : GetMajorTicks(start_ns, end_ns)) {
  const double timestamp_in_seconds = static_cast<double>(timestamp) / kNanosecondsPerSecond;

  // Number of decimal places required to display `timestamp`
  int number_of_decimal_places = 1 - static_cast<int>(std::floor(std::log10(timestamp_in_seconds))));
  max_number_of_decimal_places = std::clamp(number_of_decimal_places, max_number_of_decimal_places, kMaxDigitsPrecision);
}

return max_number_of_decimal_places;
```

Reply:
I was thinking the same and Henning beat me to it. But the calculation isn't correct, is it? `1 - log10(timestamp_in_seconds)` returns the leftmost significant digit after the comma, not the rightmost, and only works if timestamp < 1s.

Reply:
Uhh, right. Might not work. But I would still try to reverse the loops and calculcatee the number of decimal places per tick. (also due to testing as mentioned above.)

Reply:
A bit outdated, but I simply can calculate GetMajorTicks outside both loops. I wasn't optimizing at all here, but of course this doesn't hurt.

Reply:
And about the 'log change', if we were working with bits, there are a lot of tricks (I think for example (n ˆ (n-1)) + 1 should work) to clean all other bits, but I don't think we have this magic with base-10.

Reply:
We could probably see how Ryu is doing it, but that would get way too complex. :sweat_smile:  (Sorry for bringing this up in the first place. A loop is very likely the best tradeoff between complexity and readbility.)

Reply:
> how Ryu is doing it

The guy from Street Fighter is doing magic with numbers? The more you know 😄 
Anyways, I'm fine with the loop as well, it's not like this will be anywhere near performance-critical and it's easier to read and understand, so let's leave it like this.

Reply:
I rather thought about this Ryu: https://github.com/ulfjack/ryu :sweat_smile: But maybe the other one is also an option. :innocent:

Reply:
There are no relation between the methods after the fixup.